Sources/SIL/SILParser.swift

```diff
@@ -1,4 +1,21 @@
 class SILParser: Parser {
+    enum InstructionOrTerminator {
+      case instruction(Instruction)
+      case terminator(Terminator)
+    }
+
+    enum InstructionOrTerminatorDef: CustomStringConvertible {
+      case instruction(InstructionDef)
+      case terminator(TerminatorDef)
+
+      var description: String {
+        switch self {
+        case let .instruction(instructionDef): return instructionDef.description
+        case let .terminator(terminatorDef): return terminatorDef.description
+        }
+      }
+    }
```

Comment:
How do you feel about the following data model:

```
// SIL.swift
public enum InstructionDef { // formerly, InstructionOrTerminatorDef
  case nonTerminator(NonTerminatorDef)
  case terminator(TerminatorDef)
}

public enum NonTerminatorDef { // formerly, InstructionDef
  ...
}

public enum TerminatorDef {
  ...
}

public enum Instruction { // formerly, InstructionOrTerminator
  case nonTerminator(NonTerminator)
  case terminator(Terminator)
}

public enum NonTerminator { // formerly, Instruction
  ...
}

public enum Terminator {
  ...
}
```

Some aspects of the design that may be beneficial:
  1) Hopefully clearer naming (terminators are instructions as well).
  2) All data structures live in SIL.swift so that parsing and prettyprinting work the same way for them as for other data structures of the data model (both InstructionDef and Instruction will become prettyprintable via Printer, e.g. like TerminatorDef is right now).

Reply:
I feel like this will add a lot of verbosity. The only place where the instruction/terminator confusion arises is inside the parser when you never know what will come next. Your proposal propagates the same confusion throughout the whole system. For example, it does not guarantee that all defs are non-terminators and that a block ends with a terminator. Additionally, all of the pattern matching code (e.g. for abstract interpretation) now becomes a level deeper in user code, simply because it has to handle the incorrect "is a terminator" case, which should not appear in the body of the basic block anyway.

Given the added verbosity and weaker type safety I would rather keep what's there right now.

Reply:
I was in a rush when writing this comment, so I didn't explain the idea well enough. We had a brief internal chat, and the future version of this PR will reflect a similar design.